DIPs/DIP1001.md

```diff
@@ -0,0 +1,291 @@
+	TypeList:
+		Type
+		Type ,
+		Type , TypeList
+
+The throws attribute specifies that a function may throw any of the types
+specified in TypeList.
+
+@throws() is equivalent to nothrow.
+
+By default all functions are @throws(Throwable).
+
+@throws(auto) infers all thrown types
+
+functions that return auto or function templates and don't have a explict
+@throws() have @throws(auto).
+
+A function which calls a sub function with a @throws(TypeList) attribute must
+declare all uncaught exceptions within its own @throws(TypeList) attribute of
+the caller function.
+
+##### Mangling
+
+Mangling:
+
+	FuncAttr:
+		empty
+		FuncAttrPure
+		FuncAttrNogc
+		FuncAttrNothrow
+		FuncAttrProperty
+		FuncAttrRef
+		FuncAttrReturn
+		FuncAttrScope
+		FuncAttrTrusted
+		FuncAttrSafe
+		FuncAttrThrows
+	FuncAttrThrows:
+		Nt Number ThrownTypes
+	ThrownTypes:
+		Type
+		Type ThrownTypes
+
+Number is the number of elements in a @throws(TypeList) attributes, ThrownTypes
+are the list of types in TypeList.
+
+If a @throws() attribute is empty, FuncAttrNothrow should be generated instead.
+If a @throws() attribute just contains Throwable, empty should be generated
+instead.
+
+
+#### Throw Any Type
+
+Any type may be thrown not just classes that inherit from Throwable.
+
+A thrown type must be implictly convertable to one of the types mentioned in
+the @throws(TypeList) list.
+
+This fits in with the old class based exception handling.
+
+A possible optimization would be to generate a seperate unwinding table for
+each value type.
+
+#### Error Code Unwinding
+
+For all Types in a @throws(TypeList) attribute that have a manifest
+constant(__ErrorCode) and that constant evaluates to true, the compiler should
+generate C style error code handling for that function using a tagged union of
+the result and all __ErrorCode exceptions.
+
+### Rationale
+
+#### Throws Property
+
+Main Reason to include this is so Throw Any Type and Error Code Unwinding
+work with incremental compilation.
+
+Aids maintance, callers of functions can know what to catch.
+
+##### Vs C++ throws lists
+
+Due to C++'s header files and legacy ploblem, exceptions had to be checked at runtime,
+This is not the case with This proposal's @throws attribute, they can be checked at
+compile time.
+
+Unlike C++'s throws lists, @throws is usable in templates because they will be applied
+automatically.
+
+@throws being a attribute, may be applied to multiple functions at once.
+
+#### Throw Any Type
+
+Aids performace, doesn't force the exception to generate a stack trace or message.
+
+#### Error Code Unwinding
+
+Allow users to choose between the error code style of exception
+handling or no cost tables.
+
+The @throws() attribute is necessary for incremental compilation.
+
+### Breaking changes / deprecation process
+
+None
+
+### Examples
+
+Throw By Value:
+
+	struct ThrowByValue
+	{
+		int[4] filler;
+	}
+
+	int throwValue() @throws(ThrowByValue)
+	{
+		throw ThrowByValue([0, 1, 2, 3]);
```

Comment:
Throwing by value is an idea so bad only C++ could have come up with it. Where is the object going to reside ? On the stack you are currently unwinding ? It can't. So it has to live in some dynamically allocated memory by the runtime under the hood and be thrown by reference anyway.